engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/multiinstance/MultiInstanceBodyCompletedHandler.java

```diff
@@ -69,35 +73,24 @@ protected boolean handleMultiInstanceBody(
     return true;
   }
 
-  private void updateOutputCollection(
+  protected void updateOutputCollection(
       final BpmnStepContext<ExecutableMultiInstanceBody> context, final DirectBuffer variableName) {
 
     final var variablesState = context.getElementInstanceState().getVariablesState();
     final var bodyInstanceKey = context.getFlowScopeInstance().getKey();
     final var workflowKey = context.getValue().getWorkflowKey();
     final var loopCounter = context.getElementInstance().getMultiInstanceLoopCounter();
 
-    final var elementVariable = readOutputElementVariable(context);
-    final var currentCollection = variablesState.getVariableLocal(bodyInstanceKey, variableName);
-
-    final var updatedCollection = insertAt(currentCollection, loopCounter, elementVariable);
-
-    variablesState.setVariableLocal(bodyInstanceKey, workflowKey, variableName, updatedCollection);
-  }
+    // the result needs to be cloned because readOutputElementVariable(Context) uses the same buffer
+    final var currentCollection =
+        cloneBuffer(variablesState.getVariableLocal(bodyInstanceKey, variableName));
```

Comment:
We can be tricky and avoid cloning the buffer by moving `readOutputElementVariable(context);` on top :smiling_imp:  
The output element result is resolved by the expression language which uses its own buffers.